dsc-linux-4.14.18/patches-2023-01/02_winbond_flash.patch

```diff
@@ -0,0 +1,30 @@
+From eaf594b476518557e9f32a624a4c64a8b6054ad5 Mon Sep 17 00:00:00 2001
+From: David Clear <dac2@pensando.io>
+Date: Thu, 26 Jan 2023 17:24:27 -0800
+Subject: [PATCH] mtd: spi-nor: Add support for Winbond w25q02nw flash.
+
+This patch adds support for the 2Gbit Winbond w25q02nw flash,
+which supports Dual I/O and 4-byte opcodes.
+
+Signed-off-by: David Clear <dac2@pensando.io>
+---
+ drivers/mtd/spi-nor/spi-nor.c | 3 +++
+ 1 file changed, 3 insertions(+)
+
+diff --git a/drivers/mtd/spi-nor/spi-nor.c b/drivers/mtd/spi-nor/spi-nor.c
+index a0d40f7a234f..8f68f2a6dfc2 100644
+--- a/drivers/mtd/spi-nor/spi-nor.c
++++ b/drivers/mtd/spi-nor/spi-nor.c
+@@ -1167,6 +1167,9 @@ static const struct flash_info spi_nor_ids[] = {
+ 	{ "w25q256", INFO(0xef4019, 0, 64 * 1024, 512, SECT_4K | SPI_NOR_DUAL_READ | SPI_NOR_QUAD_READ) },
+ 	{ "w25m512jv", INFO(0xef7119, 0, 64 * 1024, 1024,
+ 			SECT_4K | SPI_NOR_QUAD_READ | SPI_NOR_DUAL_READ) },
++	{ "w25q02nw", INFO(0xef8022, 0, 64 * 1024, 4096,
++			    SECT_4K | SPI_NOR_DUAL_READ | SPI_NOR_QUAD_READ |
++			    SPI_NOR_4B_OPCODES) },
```

Comment:
Any idea why we need this? We did not have this option added to 5.4.173 patch.

Reply:
The only difference here is SPI_NOR_4B_OPCODES.
The newer kernels will automatically use 4B opcodes for these flashes.  4.14 won't, for Winbond, unless this flag is turned on here.
I was looking for a Micron Flash Elba yesterday because I tested the whole matrix of 5.10, 5.4, 4.14 on Micron, Macronix and Winbond flashes to verify which flash opcodes would be used.

Without this flag, 4.14 will issue a command to put the Winbond into "4-byte mode".  What this does is it lets the controller use 3-byte instructions (e.g. command 0x02 for write), yet take 4 address bytes.

This makes me nervous as, if the command fails to land (let's just say that's highly unlikely, but whatever) then the device will incorrectly interpret the transaction data.  I don't want there to be any chance of accidentally corrupting the eSecure flash due to mis-parsed address field.

So I opt to make sure that 4B opcodes (e.g. 0x12 for write) are used always.

Reply:
I see. Got it.